### PR TITLE
Track file access for dune-project and opam files.

### PR DESCRIPTION
**Pull request by rgrinberg**

### CODE REVIEW

**Comment by rgrinberg** [comment]
@snowleopard this one goes along with #5652

**Comment by rgrinberg** [comment]
@snowleopard this one goes along with #5652

File: src/dune_engine/dune_project.ml [hunk 1]
```diff
@@ -762,219 +764,236 @@ let parse ~dir ~lang ~opam_packages ~file ~dir_status =
...
     String.Set.fold files ~init:[] ~f:(fun fn acc ->
         match Package.Name.of_opam_file_basename fn with
         | None -> acc
         | Some name ->
           let opam_file = Path.Source.relative dir fn in
           let loc = Loc.in_file (Path.source opam_file) in
-          (* We are lazy here because if the opam files are generated by dune,
-             there is no point parsing them. This value will only be forced if
-             generation of opam files is disabled. *)
-          let pkg = lazy (Package.load_opam_file opam_file name) in
```
Comment (snowleopard) [inline src/dune_engine/dune_project.ml:964]: Why did you remove the laziness? Do you think it wasn't useful?

File: src/dune_engine/opam_file.mli [hunk 2]
```diff
@@ -7,7 +7,9 @@ open OpamParserTypes
 type t = opamfile
 
 (** Load a file *)
-val load : Path.t -> t
+val load : Path.t -> (t, exn) result Memo.t
+
+val load_untracked : Path.t -> t
```
Comment (snowleopard) [inline src/dune_engine/opam_file.mli:12]: Shall we keep the convention of using `Untracked` modules?

Also, the type suggests that it's not only untracked but can't fail. Is that true?

**Review by snowleopard — COMMENTED** [review]
Generally looks great, but I left some questions.

File: src/dune_engine/opam_file.mli (hunk 2 above)
Comment (rgrinberg) [reply src/dune_engine/opam_file.mli:12]: Tbh, I'm not a fan of the Untracked convention here because the return type already provides this information. I did the following: inline the IO functions in the callers. It makes both the error handling and tracked vs. untracked more clear.

File: src/dune_engine/dune_project.ml (hunk 1 above)
Comment (rgrinberg) [reply src/dune_engine/dune_project.ml:964]: I think the work is still delayed because we aren't binding on the return Memo.t value here. It is left implicit that we are going to be binding on them only once later though. I could add back the lazy if you think it's more clear.

Comment (snowleopard) [reply src/dune_engine/dune_project.ml:964]: Oh, you are right. Sorry.

**Review by snowleopard — APPROVED** [review]
Thanks!
